index-configuration-tool/main.py

```diff
@@ -0,0 +1,135 @@
+import sys
+import requests
+import logstash_conf_parser as parser
+from jsondiff import diff
+
+# Constants
+SUPPORTED_ENDPOINTS = ["opensearch", "elasticsearch"]
+HOSTS_KEY = "hosts"
+USER_KEY = "user"
+PWD_KEY = "password"
+SETTINGS_KEY = "settings"
+MAPPINGS_KEY = "mappings"
+INDEX_KEY = "index"
+INTERNAL_SETTINGS_KEYS = ["creation_date", "uuid", "provided_name", "version", "store"]
+ALL_INDICES_ENDPOINT = "*"
+
+
+# Utility method to make a comma-separated string from a set
+def string_from_set(s: set) -> str:
+    if s:
+        return ", ".join(s)
+    else:
+        return "[]"
+
+
+# Utility method to get the plugin config from the base data
+def get_plugin_config(data: dict, plugin_type: str) -> dict:
+    # Tuple contents are (type, config)
+    return data[plugin_type][1]
+
+
+# TODO Only supports basic auth for now
+def get_auth(input_data: dict) -> tuple:
+    if USER_KEY in input_data and PWD_KEY in input_data:
+        return input_data[USER_KEY], input_data[PWD_KEY]
+
+
+def get_endpoint_info(plugin_config: dict) -> tuple:
+    endpoint = "https://" if ("ssl" in plugin_config and plugin_config["ssl"]) else "http://"
+    endpoint += plugin_config[HOSTS_KEY][0] if type(plugin_config[HOSTS_KEY]) is list else plugin_config[HOSTS_KEY]
```

Comment:
The `hosts` key for both the OS and [ES plugins](https://www.elastic.co/guide/en/logstash/7.10/plugins-inputs-elasticsearch.html#plugins-inputs-elasticsearch-hosts) could either be a simple string (in the case of a single host/endpoint) or an array of hosts (in which case Logstash round-robins through them per `bulk` request).

For this tool, we don't need to support round-robin since we're just creating indices and not ingesting data. Thus, the code here simply picks the first value if an array is provided. I've added the following comment to the code:

```
# "hosts" can be a simple string, or an array of hosts for Logstash to hit.
# This tool needs one accessible host, so we pick the first entry in the latter case.
```